src/main/java/com/github/dockerjava/core/command/PullImageResultCallback.java

```diff
@@ -19,14 +21,71 @@
 
     private final static Logger LOGGER = LoggerFactory.getLogger(PullImageResultCallback.class);
 
+    private boolean isSwarm = false;
+    private Map<String, PullResponseItem> results = null;
     private PullResponseItem latestItem = null;
 
     @Override
     public void onNext(PullResponseItem item) {
-        this.latestItem = item;
+        // only do it once
+        if (results == null && latestItem == null) {
+            checkForDockerSwarmResponse(item);
+        }
+
+        if (isSwarm) {
+            handleDockerSwarmResponse(item);
+        } else {
+            handleDockerClientResponse(item);
+        }
         LOGGER.debug(item.toString());
     }
 
+    public void checkForDockerSwarmResponse(PullResponseItem item) {
+        if (item.getStatus().matches("Pulling\\s.+\\.{3}$")) {
+            isSwarm = true;
+            LOGGER.debug("Communicating with Docker Swarm.");
+        }
+    }
+
+    public void handleDockerSwarmResponse(PullResponseItem item) {
+        if (results == null) {
+            results = new HashMap<String, PullResponseItem>();
+        }
+        results.put(item.getId(), item);
+    }
+
+    public void checkDockerSwarmPullSuccessful() {
+        if (results.isEmpty()) {
+            throw new DockerClientException("Could not pull image");
+        } else {
+            boolean pullFailed = false;
+            StringBuilder sb = new StringBuilder();
+
+            for (PullResponseItem pullResponseItem : results.values()) {
```

Comment:
Swarm works like so. When you issue a pull cmd, it returns following responses as PullResponseItem for each node in the swarm:

21:38:57.269 DEBUG c.g.d.c.c.PullImageResultCallback - PullResponseItem[stream=<null>,status=Pulling ci1.camunda.loc:5000/camunda-ci-postgresql:9.4...,progressDetail=ResponseItem.ProgressDetail[current=0,total=0,start=0],progress=<null>,id=ci7.camunda.loc,from=<null>,time=0,errorDetail=<null>,error=<null>]

When a pull is successfully for a swarm node following PullResponseItem is returned:

21:39:04.513 DEBUG c.g.d.c.c.PullImageResultCallback - PullResponseItem[stream=<null>,status=Pulling ci1.camunda.loc:5000/camunda-ci-postgresql:9.4... : downloaded,progressDetail=ResponseItem.ProgressDetail[current=0,total=0,start=0],progress=<null>,id=ci3.camunda.loc,from=<null>,time=0,errorDetail=<null>,error=<null>]

My code will first collect all PullResponseItems ordered by node id as key. So only the latest PullResponseItem of each node will be stored, previous ones overwritten.
When the PullImageCmd is finished and awaitSuccess method continues to evaluate the success of the pull, it will check the last result of each node. When one of those results indicates a failed pull, it will throw an exception with all nodes where the pull failed.